compiler/rustc_infer/src/infer/combine.rs

```diff
@@ -532,6 +545,11 @@ impl<'tcx> TypeRelation<'tcx> for Generalizer<'_, 'tcx> {
         true
     }
 
+    fn mark_ambiguous(&mut self) {
+        // The generalizer always compares types against themselves,
+        // and thus doesn't really take part in coherence.
+    }
```

Comment:
that is interesting :thinking: 

if we relate `?0 = impl Trait<?1>` during coherence, we end up generalizing `impl Trait<?1>` to instantiate `?0` which results in using `super_relate_tys` for `a: impl Trait<?1>`, `b: impl Trait<?1>`. This then keeps `?1` so we end up setting `?0` to `impl Trait<?1>` instead of `impl Trait<?2>`.

This isn't really an issue, except that `?1` may be an inference variable from a higher universe than `?0` which would mean that `?0` would be allowed to name stuff from that universe "through" `?1`, causing us to not detect a higher ranked mismatch.

I have no idea how to even start writing a test case for that though.